pallets/system/src/header_builder.rs

```diff
@@ -104,51 +108,129 @@ fn corrupt_commitment(block_number: u32, commitment: &mut Vec<u8>) {
 		app_extrinsics.to_vec(),
 		MIN_WIDTH,
 		block_length.cols.0.saturated_into(), // even if we run on a u16 target this is fine
 		block_length.rows.0.saturated_into(),
 		seed,
 	)
-	.expect("Grid construction cannot fail");
+	.map_err(|e| format!("Grid construction failed: {e:?}"))?;
 
-	// Evaluation Grid Build Time Metrics
-	HeaderExtensionBuilderMetrics::observe_evaluation_grid_build_time(timer.elapsed());
+	Ok(grid)
+}
 
-	let timer = std::time::Instant::now();
-	let commitment = grid
+#[cfg(feature = "std")]
+pub fn build_commitment(grid: &EvaluationGrid) -> Result<Vec<u8>, String> {
+	use kate::gridgen::AsBytes;
+	use once_cell::sync::Lazy;
+
+	// couscous has pp for degree upto 1024
+	static PMP: Lazy<kate::pmp::m1_blst::M1NoPrecomp> =
+		Lazy::new(kate::couscous::multiproof_params);
+
+	let poly_grid = grid
 		.make_polynomial_grid()
-		.expect("Make polynomials cannot fail")
+		.map_err(|e| format!("Make polynomial grid failed: {e:?}"))?;
+
+	let extended_grid = poly_grid
 		.extended_commitments(&*PMP, 2)
-		.expect("Extended commitments cannot fail")
+		.map_err(|e| format!("Grid extension failed: {e:?}"))?;
+
+	let commitment = extended_grid
 		.iter()
 		.flat_map(|c| c.to_bytes().expect("Commitment serialization cannot fail"))
 		.collect::<Vec<u8>>();
 
+	Ok(commitment)
+}
+
+#[cfg(feature = "std")]
+pub fn get_empty_header(data_root: H256, version: HeaderVersion) -> HeaderExtension {
+	use avail_core::header::extension::{v1, v2};
+	let empty_commitment: Vec<u8> = vec![0];
+	let empty_app_lookup = DataLookup::new_empty();
+
+	match version {
+		HeaderVersion::V1 => {
+			use avail_core::kate_commitment::v1::KateCommitment;
+			let kate = KateCommitment {
+				rows: 1,
+				cols: 4,
+				commitment: empty_commitment,
+				data_root,
+			};
+			v1::HeaderExtension {
+				app_lookup: empty_app_lookup,
+				commitment: kate,
+			}
+			.into()
+		},
+		HeaderVersion::V2 => {
+			use avail_core::kate_commitment::v2::KateCommitment;
+			let kate = KateCommitment::new(1, 4, data_root, empty_commitment);
+			v2::HeaderExtension {
+				app_lookup: empty_app_lookup,
+				commitment: kate,
+			}
+			.into()
+		},
+	}
+}
+
+#[cfg(feature = "std")]
+pub fn build_extension(
+	app_extrinsics: &[AppExtrinsic],
+	data_root: H256,
+	block_length: BlockLength,
+	_block_number: u32,
+	seed: Seed,
+	version: HeaderVersion,
+) -> HeaderExtension {
+	use avail_base::metrics::avail::HeaderExtensionBuilderMetrics;
+	use avail_core::header::extension::{v1, v2};
+
+	let build_extension_start = std::time::Instant::now();
+
+	// Build the grid
+	let timer = std::time::Instant::now();
+	let maybe_grid = build_grid(app_extrinsics, block_length, seed);
+	// Evaluation Grid Build Time Metrics
+	HeaderExtensionBuilderMetrics::observe_evaluation_grid_build_time(timer.elapsed());
+	let grid = match maybe_grid {
+		Ok(res) => res,
+		Err(message) => {
+			log::error!("NODE_CRITICAL_ERROR_001 - A critical error has occured: {message:?}.");
```

Comment:
Can we have some other error code? Looks verbose. Something like this (although not equivalent)?
https://github.com/availproject/avail/wiki/Transaction-Custom-IDs

Reply:
And then a corresponding error on the wiki,

Reply:
We chose to not introduce the same type of error for multiple reasons : 
- This is a console only error
- This is not a transaction error that need to be returned to the user so we cannot use the same feat
- It needs to be easily findable